### DISCUSSION TITLE
How can we mark partitions as "materialized" in dagster if the data already exists in the DWH?

### DISCUSSION THREAD

**Discussion by cyberjar09** [dagster-io/dagster#17847] [Q&A] [answered]
Is there a way to tell Dagster that the first thousand partitions of a time-partitioned asset are already completed ? If I copy data manually into a table?

I'd like to mark these assets and partitions as materialized without actually running the code that overwrites their contents.

### Context:
- our self hosted dagster instance is undergoing a major overhaul
  - we are migrating our dagster webserver+daemon deployment to another kubernetes cluster
  - this new daemon+webserver will also be connected to multiple code locations which have clashing names (dagster assets must be uniquely named across code locations)
    - hence we need to also rename all assets in the code locations to make them unique (we are adding a region prefix)

---
### Problem Statement:
- given the above, we will lose all our dagster run and materialization history
  - asset rename 
  - moving dagster to actually consolidate multiple server instances deployed across k8s clusters to 1 instance that just has multiple code locations instead
- since the historical data already exists in the DWH, we wanted the ability to just mark those partitions as "materialized" rather than re-run all the backfills just to get the UI to reflect that the data exists
- one idea was to just mark the date range and do a no_op in dagster code but was wondering if there are alternatives to consider?

**Comment by sryza** [comment, marked as answer]
You can report assets and asset partitions as materialized in Dagster without actually running the computation that materializes. This is possible from the UI and using Dagster's Python APIs.

### From the UI

If you go to the page for the asset, you can click the dropdown next to the "Materialize" button and click "Report materialization events".

<img width="1154" alt="image" src="https://github.com/dagster-io/dagster/assets/654855/635c135f-c2ca-4ddf-9f92-ed6785d4426e">

Then you'll get a dialog that allows you to choose a partition range:

<img width="711" alt="image" src="https://github.com/dagster-io/dagster/assets/654855/f4fed3fc-9e57-4d1f-bdc0-cb4bfab5d8a3">

### Using the Python APIs

First, get your hands on a `DagsterInstance`: https://github.com/dagster-io/dagster/discussions/7427.

Then you can use `DagsterInstance.report_runless_asset_event`:

```python
from dagster import AssetMaterialization, AssetKey

for partition_key in partition_keys:
    asset_materialization = AssetMaterialization(AssetKey("my_asset_key"), partition=partition_key)
    instance.report_runless_asset_event(asset_materialization)
```

**Reply by weberdavid** [reply]
@sryza Thanks. Two questions:

- Would this work with DynamicPartitionDefinitions?
- Is there a way to do it programatically in Python? Let's say I could write a function that gets all the partition-keys out of my database, and based on that I would want to fill those keys as already materialized partitions in Dagster.

**Reply by sryza** [reply]
* This should work with DynamicPartitionDefinitions
* Yes. I updated the answer above to include it.

We also have an issue to track adding CLI support: https://github.com/dagster-io/dagster/issues/17202

**Reply by JasperHG90** [reply]
Hi @sryza . Stumbled upon this github issue. I'm on Dagsterr 1.6.5. and was looking for this functionality but it's not available. Do you know if this has been moved?

**Reply by sryza** [reply]
Are you able to send a screenshot. I just checked on the asset details page and saw it there.

**Reply by JasperHG90** [reply]
I found it, thank you!

**Comment by cyberjar09** [comment]
thanks a lot for the detailed response @sryza 
wanted to confirm that this feature is only available 1.5.3 onwards yea?

**Reply by sryza** [reply]
That's correct

**Comment by dsgibbons** [comment]
Is there a way to do the opposite? That is, report that a previously materialized partition is now failed or missing? Both via the UI and the Python API?

**Reply by sryza** [reply]
Not currently. Here's the relevant issue: https://github.com/dagster-io/dagster/issues/14749.

**Comment by varunp2k** [comment]


**Reply by varunp2k** [reply]
@sryza 
can you please help?

**Comment by eliot1785** [comment]
One follow-up question would be, how exactly could you run that Python code? From here, you can see that any Python code which runs DagsterInstance.get() would need to have access to either a dagster.yaml or to actually be running inside a Code Location.

However if you are running Dagster in the cloud that gets tricky, because whatever dagster.yaml you have locally is likely just for some toy Dagster which isn't where you are trying to add the partitions. Other than creating a fake asset/op just for the purposes of creating these pre-existing partitions, I'm not sure how you'd gain access to the right Dagster server.
